Yank/restraints.py

```diff
@@ -1564,46 +1737,73 @@ def _pick_restrained_atoms(self, sampler_state, topography):
         Future updates can further refine this algorithm.
 
         """
-        t = md.Trajectory(sampler_state.positions / unit.nanometers, topography.topology)
+        # No need to determine parameters if atoms have been given.
+        if self.restrained_receptor_atoms is not None and self.restrained_ligand_atoms is not None:
+            return self.restrained_receptor_atoms + self.restrained_ligand_atoms
+
+        # If receptor and ligand atoms are explicitly provided, use those.
+        heavy_ligand_atoms = self.restrained_ligand_atoms
+        heavy_receptor_atoms = self.restrained_receptor_atoms
+
+        # Otherwise we restrain only heavy atoms.
+        heavy_atoms = set(topography.topology.select('not element H').tolist())
+        # Intersect heavy atoms with receptor/ligand atoms (s1&s2 is intersect).
+        if heavy_ligand_atoms is None:
```

Comment:
Just noticed, if user chosen, hydrogens can get selected as the heavy atoms to use here, may be a problem.